src/components/pages/devices/flyouts/deviceJobs/deviceJobTags.js

```diff
@@ -1,20 +1,304 @@
+    this.jobNameLink = this.linkTo('jobName')
+      .reject(nonAlphaNumeric)
+      .check(Validator.notEmpty, () => this.props.t('devices.flyouts.jobs.validation.required'));
+
+    this.tagsLink = this.linkTo('commonTags');
+  }
+
+  componentDidMount() {
+    if (this.props.devices) {
+      this.populateState(this.props.devices);
+    }
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.devices && (this.props.devices || []).length !== nextProps.devices.length) {
+      this.populateState(nextProps.devices);
+    }
+  }
+
+  populateState(devices) {
+    // Create a stream of devices
+    const deviceStream = Observable.of(devices).flatMap(_ => _);
+    deviceStream
+      // Convert the device stream into a stream of tab names
+      .flatMap(device => Object.keys(device.tags))
+      // Remove duplicate tag names from the stream
+      .distinct()
+      // Extract only the common tag names
+      .filter(tagName =>
+        devices.every(device => (typeof (device.tags[tagName]) !== 'undefined'))
+      )
+      // Compute the tag values for each tag name into a map
+      .flatMap(tagName =>
+        deviceStream
+          .map(device => {
+            // Explicitly check undefined because empty string is valid
+            return device.tags[tagName];
+          })
+          .distinct()
+          .filter(value => value !== undefined)
+          .reduce((acc, curr) => [...acc, curr], [])
+          .map(values => ({ tagName: tagName, values: values }))
+      )
+      // Construct the new component state
+      .reduce(
+        (newState, { tagName, values }) => {
+          const value = values.length === 1 ? values[0] : TagJobConstants.multipleValues;
+          const type = values.every(isNumeric) ? TagJobConstants.numberType : TagJobConstants.stringType;
+          return (update(newState, {
+            commonTags: { $push: [{ name: tagName, value, type }] }
+          }));
+        },
+        initialState
+      )
+      // Update the component state
+      .subscribe(newState => {
+        this.setState(newState);
+      });
+  }
+
+  formIsValid() {
+    return [
+      this.jobNameLink
+    ].every(link => !link.error);
+  }
+
+  apply = (event) => {
+    event.preventDefault();
+    if (this.formIsValid()) {
+      this.setState({ isPending: true });
+
+      const { devices } = this.props;
+      const { commonTags, deletedTags } = this.state;
+      const request = toSubmitTagsJobRequestModel(devices, this.state);
+
+      this.subscription = IoTHubManagerService.sumbitJob(request)
+        .subscribe(
+          ({ jobId }) => {
+            this.setState({ jobId, successCount: devices.length, isPending: false, changesApplied: true });
+            this.props.updateTags({ deviceId: devices.map(({ id }) => id), commonTags, deletedTags });
+          },
+          errorResponse => {
+            this.setState({ error: errorResponse.errorMessage, isPending: false, changesApplied: true });
```

Comment:
For handling the error. You can save it directly and use the `AjaxError` component I made and placed in shared. It has a unified error formatting and handles all the error translation stuff. You can find examples in the dashboard panels